--- a/drivers/gpu/drm/panel/panel-ilitek-ili9881c.c	2023-07-01 20:46:40.000000000 +0000
+++ b/drivers/gpu/drm/panel/panel-ilitek-ili9881c.c	2023-08-08 23:47:37.069526508 +0000
@@ -919,7 +919,10 @@
 	dsi->format = MIPI_DSI_FMT_RGB888;
 	dsi->lanes = 4;
 
-	return mipi_dsi_attach(dsi);
+	ret =  mipi_dsi_attach(dsi);
+	if (ret)
+		drm_panel_remove(&ctx->panel);
+	return ret;
 }
 
 static void ili9881c_dsi_remove(struct mipi_dsi_device *dsi)
