From a2c74052440aaec196ac39dbfd48dd7746b7f888 Mon Sep 17 00:00:00 2001
From: ArenM <aren@peacevolution.org>
Date: Sat, 29 Jan 2022 20:54:15 -0500
Subject: [PATCH sxmo-utils] Let systemd handle service supervision

---
 scripts/modem/sxmo_modem.sh | 33 +++------------------------------
 1 file changed, 3 insertions(+), 30 deletions(-)

diff --git a/scripts/modem/sxmo_modem.sh b/scripts/modem/sxmo_modem.sh
index 752029c..d11354b 100644
--- a/scripts/modem/sxmo_modem.sh
+++ b/scripts/modem/sxmo_modem.sh
@@ -154,20 +154,7 @@ checkfornewtexts() {
 		if [ -z "$TEXTDATA" ]; then
 			stderr "Received an empty SMS (TEXTID: $TEXTID).  I will assume this is an MMS."
 			printf %b "$(date +%FT%H:%M:%S%z)\tdebug_mms\tNULL\tEMPTY (TEXTID: $TEXTID)\n" >> "$SXMO_LOGDIR/modemlog.tsv"
-			if [ -f "${SXMO_MMS_BASE_DIR:-"$HOME"/.mms/modemmanager}/mms" ]; then 
-				if sxmo_daemons.sh running mmsd -q; then
-					continue
-				else
-					stderr "mmsd not running."
-					if pgrep -f sxmo_mmsdconfig.sh >/dev/null; then
-						stderr "mmsdconfig running."
-						continue
-					fi
-					stderr "restarting mmsd."
-					sxmo_daemons.sh start mmsd mmsdtng "$SXMO_MMSD_ARGS"
-					continue
-				fi
-			else
+			if [ ! -f "${SXMO_MMS_BASE_DIR:-"$HOME"/.mms/modemmanager}/mms" ]; then 
 				stderr "WARNING: mmsdtng not found or unconfigured, treating as normal sms."
 			fi
 		fi
@@ -188,8 +175,7 @@ checkfornewtexts() {
 				if pgrep -f sxmo_vvmdconfig.sh >/dev/null; then
 					stderr "vvmdconfig running, doing nothing."
 				else
-					stderr "starting vvmd."
-					sxmo_daemons.sh start vvmd vvmd "$SXMO_VVMD_ARGS"
+					stderr "WARNING: vvmd isn't running"
 				fi
 			fi
 		fi
@@ -215,20 +201,7 @@ checkfornewtexts() {
 		if [ "$TEXT" = "--" ]; then
 			stderr "Text from $NUM (TEXTID: $TEXTID) with '--'.  I will assume this is an MMS."
 			printf %b "$TIME\tdebug_mms\t$NUM\t$TEXT\n" >> "$SXMO_LOGDIR/modemlog.tsv"
-			if [ -f "${SXMO_MMS_BASE_DIR:-"$HOME"/.mms/modemmanager}/mms" ]; then
-				if sxmo_daemons.sh running mmsd -q; then
-					continue
-				else
-					stderr "mmsd not running."
-					if pgrep -f sxmo_mmsdconfig.sh >/dev/null; then
-						stderr "mmsdconfig running."
-						continue
-					fi
-					stderr "restarting mmsd."
-					sxmo_daemons.sh start mmsd mmsdtng "$SXMO_MMSD_ARGS"
-					continue
-				fi
-			else
+			if [ ! -f "${SXMO_MMS_BASE_DIR:-"$HOME"/.mms/modemmanager}/mms" ]; then
 				stderr "WARNING: mmsdtng not found or unconfigured, treating as normal sms."
 			fi
 		fi
-- 
2.35.1

