diff --git a/Telegram/SourceFiles/window/window.style b/Telegram/SourceFiles/window/window.style
index 854c5df4b..66ded5cc5 100644
--- a/Telegram/SourceFiles/window/window.style
+++ b/Telegram/SourceFiles/window/window.style
@@ -10,8 +10,8 @@ using "ui/widgets/widgets.style";
 using "ui/chat/chat.style";
 using "boxes/boxes.style"; // UserpicButton
 
-windowMinWidth: 380px;
-windowMinHeight: 480px;
+windowMinWidth: 300px;
+windowMinHeight: 400px;
 windowDefaultWidth: 800px;
 windowDefaultHeight: 600px;
 windowBigDefaultWidth: 1024px;
@@ -20,7 +20,7 @@ windowResizeArea: 4px;
 
 columnMinimalWidthLeft: 260px;
 columnMaximalWidthLeft: 540px;
-columnMinimalWidthMain: 380px;
+columnMinimalWidthMain: 300px;
 columnDesiredWidthMain: 512px;
 columnMinimalWidthThird: 292px;
 columnMaximalWidthThird: 392px;
