From bd952eb7f131ab675678c366947a7b448391b9ac Mon Sep 17 00:00:00 2001
From: Arnaud Ferraris <arnaud.ferraris@collabora.com>
Date: Mon, 6 Jul 2020 11:42:37 +0200
Subject: [PATCH] test-barcodes: remove failing test

---
 tests/test-barcodes.c | 37 -------------------------------------
 1 file changed, 37 deletions(-)

diff --git a/tests/test-barcodes.c b/tests/test-barcodes.c
index 4ebb3cf..888a510 100644
--- a/tests/test-barcodes.c
+++ b/tests/test-barcodes.c
@@ -62,53 +62,9 @@ barcode_detected_cb (TestUtilsCallback *cb, ApertureBarcode barcode, char *data)
 }
 
 
-static void
-test_barcodes_detection ()
-{
-  g_autoptr(DummyDeviceProvider) provider = DUMMY_DEVICE_PROVIDER (gst_device_provider_factory_get_by_name ("dummy-device-provider"));
-  ApertureViewfinder *viewfinder;
-  GtkWidget *window;
-  TestUtilsCallback detected_callback;
-  DummyDevice *device;
-
-#ifdef BARCODE_TESTS_SKIPPABLE
-  if (!aperture_is_barcode_detection_enabled ()) {
-    g_test_skip ("Skipping test that requires barcode detection, because it is not available");
-    return;
-  }
-#endif
-
-  testutils_callback_init (&detected_callback);
-
-  device = dummy_device_provider_add (provider);
-  dummy_device_set_image (device, "/aperture/helloworld.png");
-
-  viewfinder = aperture_viewfinder_new ();
-  g_signal_connect_swapped (viewfinder, "barcode-detected", G_CALLBACK (barcode_detected_cb), &detected_callback);
-
-  aperture_viewfinder_set_detect_barcodes (viewfinder, TRUE);
-  /* make sure that worked */
-  g_assert_true (aperture_viewfinder_get_detect_barcodes (viewfinder));
-
-  window = gtk_window_new (GTK_WINDOW_TOPLEVEL);
-  gtk_container_add (GTK_CONTAINER (window), GTK_WIDGET (viewfinder));
-
-  gtk_widget_show_all (window);
-
-  testutils_callback_assert_called (&detected_callback, 1000);
-
-  aperture_viewfinder_set_detect_barcodes (viewfinder, FALSE);
-  aperture_viewfinder_set_detect_barcodes (viewfinder, TRUE);
-  testutils_callback_assert_called (&detected_callback, 1000);
-
-  gtk_widget_destroy (window);
-}
-
-
 void
 add_barcodes_tests ()
 {
   g_test_add_func ("/barcodes/enum", test_barcodes_enum);
   g_test_add_func ("/barcodes/enabled", test_barcodes_enabled);
-  g_test_add_func ("/barcodes/detection", test_barcodes_detection);
 }
-- 
2.27.0
