From f7dbdd239b5eb86374c96b2cb0de0c46fccafaf1 Mon Sep 17 00:00:00 2001
From: Mohammed Sadiq <sadiq@sadiqpk.org>
Date: Thu, 27 Aug 2020 14:48:11 +0530
Subject: [PATCH 15/29] gtkprivate: Add an API to check if phone

---
 gtk/gtkprivate.h | 21 +++++++++++++++++++++
 1 file changed, 21 insertions(+)

diff --git a/gtk/gtkprivate.h b/gtk/gtkprivate.h
index 9c0a35a..3aa34cb 100644
--- a/gtk/gtkprivate.h
+++ b/gtk/gtkprivate.h
@@ -126,6 +126,27 @@ guint gtk_get_display_debug_flags (GdkDisplay *display);
 
 #endif /* G_ENABLE_DEBUG */
 
+static inline gboolean
+_gtk_get_is_phone (void)
+{
+  GSettings *gsettings;
+  GSettingsSchema *schema;
+  gboolean is_phone;
+
+  schema = g_settings_schema_source_lookup (g_settings_schema_source_get_default (),
+                                            "org.gtk.Settings.Purism", TRUE);
+
+  if (!schema)
+    return FALSE;
+
+  gsettings = g_settings_new_full (schema, NULL, NULL);
+  is_phone = g_settings_get_boolean (gsettings, "is-phone");
+  g_object_unref (gsettings);
+  g_settings_schema_unref (schema);
+
+  return is_phone;
+}
+
 G_END_DECLS
 
 #endif /* __GTK_PRIVATE_H__ */
-- 
2.29.2

