From 22ac2c6fecc98efd11d96dc3e04938b8777a3510 Mon Sep 17 00:00:00 2001
From: Rico Tzschichholz <ricotz@ubuntu.com>
Date: Thu, 2 Jul 2020 15:34:42 +0200
Subject: [PATCH] Don't use unsupported vala syntax

---
 src/contacts-avatar-utils.vala | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/contacts-avatar-utils.vala b/src/contacts-avatar-utils.vala
index 15e84164..c5c6d61d 100644
--- a/src/contacts-avatar-utils.vala
+++ b/src/contacts-avatar-utils.vala
@@ -126,7 +126,7 @@ namespace Contacts.AvatarUtils {
 
   public Gdk.RGBA get_color_for_name(string name) {
     // https://gitlab.gnome.org/Community/Design/HIG-app-icons/blob/master/GNOME%20HIG.gpl
-    const double[,3] GNOME_COLOR_PALETTE = {
+    const double[,] GNOME_COLOR_PALETTE = {
       {  98, 160, 234 },
       {  53, 132, 228 },
       {  28, 113, 216 },
-- 
GitLab

