From 781c122dfa0c9ae67ec00aa9aece93592a638f3b Mon Sep 17 00:00:00 2001
From: Mohammed Sadiq <sadiq@sadiqpk.org>
Date: Thu, 19 Sep 2019 20:44:12 +0530
Subject: [PATCH 2/4] wifi: Use new connection editor

---
 panels/network/net-device-wifi.c | 2 ++
 1 file changed, 2 insertions(+)

Index: gnome-control-center/panels/network/net-device-wifi.c
===================================================================
--- gnome-control-center.orig/panels/network/net-device-wifi.c	2021-10-23 07:56:40.314000000 +0000
+++ gnome-control-center/panels/network/net-device-wifi.c	2021-10-23 07:56:40.311000000 +0000
@@ -962,6 +962,8 @@
         return 0;
 }
 
+#include "connection-editor/cc-connection-editor.h"
+#include "pureos.h"
 static void
 show_details_for_row (NetDeviceWifi *self, CcWifiConnectionRow *row, CcWifiConnectionList *list)
 {
@@ -972,6 +974,7 @@
         connection = cc_wifi_connection_row_get_connection (row);
         ap = cc_wifi_connection_row_best_access_point (row);
 
+#include "use-new-editor.c"
         editor = net_connection_editor_new (connection, self->device, ap, self->client);
         gtk_window_set_transient_for (GTK_WINDOW (editor), GTK_WINDOW (gtk_widget_get_toplevel (GTK_WIDGET (row))));
         gtk_window_present (GTK_WINDOW (editor));
