From: Arnaud Ferraris <arnaud.ferraris@collabora.com>
Date: Mon, 25 May 2020 21:03:38 +0200
Subject: [PATCH 4/4] power: add more suspend timing options

---
 panels/power/cc-power-panel.ui | 8 ++++++++
 1 file changed, 8 insertions(+)

Index: gnome-control-center/panels/power/cc-power-panel.ui
===================================================================
--- gnome-control-center.orig/panels/power/cc-power-panel.ui	2022-03-31 05:05:48.557000000 +0000
+++ gnome-control-center/panels/power/cc-power-panel.ui	2022-03-31 05:05:48.553000000 +0000
@@ -9,6 +9,14 @@
     </columns>
     <data>
       <row>
+        <col id="0" translatable="yes" context="automatic_suspend" comments="Translators: Option for &quot;Delay&quot; in &quot;Automatic suspend&quot; dialog.">5 minutes</col>
+        <col id="1">300</col>
+      </row>
+      <row>
+        <col id="0" translatable="yes" context="automatic_suspend" comments="Translators: Option for &quot;Delay&quot; in &quot;Automatic suspend&quot; dialog.">10 minutes</col>
+        <col id="1">600</col>
+      </row>
+      <row>
         <col id="0" translatable="yes" context="automatic_suspend" comments="Translators: Option for &quot;Delay&quot; in &quot;Automatic suspend&quot; dialog.">15 minutes</col>
         <col id="1">900</col>
       </row>
