Index: gnome-control-center/panels/keyboard/cc-keyboard-panel.ui
===================================================================
--- gnome-control-center.orig/panels/keyboard/cc-keyboard-panel.ui	2021-10-23 07:56:40.757000000 +0000
+++ gnome-control-center/panels/keyboard/cc-keyboard-panel.ui	2021-10-23 07:56:40.754000000 +0000
@@ -128,6 +128,7 @@
                                 <property name="visible">True</property>
                                 <property name="margin_left">6</property>
                                 <property name="use_underline">True</property>
+                                <property name="wrap">True</property>
                                 <property name="label" translatable="yes">Use the _same source for all windows</property>
                               </object>
                             </child>
@@ -155,6 +156,7 @@
                                 <property name="visible">True</property>
                                 <property name="margin_left">6</property>
                                 <property name="use_underline">True</property>
+                                <property name="wrap">True</property>
                                 <property name="label" translatable="yes">Switch input sources _individually for each window</property>
                               </object>
                             </child>
