From e390b8e76f796e5b1d619cf36a0ed46103029e0b Mon Sep 17 00:00:00 2001
From: Mohammed Sadiq <sadiq@sadiqpk.org>
Date: Wed, 25 Sep 2019 21:27:13 +0530
Subject: [PATCH 3/8] notification: Hide lockscreen notification option on
 phones

---
 panels/notifications/cc-notifications-panel.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/panels/notifications/cc-notifications-panel.c b/panels/notifications/cc-notifications-panel.c
index 510d7065d..2eee23a2c 100644
--- a/panels/notifications/cc-notifications-panel.c
+++ b/panels/notifications/cc-notifications-panel.c
@@ -212,6 +212,9 @@ cc_notifications_panel_init (CcNotificationsPanel *panel)
                             cc_panel_get_cancellable (CC_PANEL (panel)),
                             on_perm_store_ready,
                             panel);
+
+  if (pureos_get_is_phone ())
+    gtk_widget_hide (GTK_WIDGET (panel->lock_screen_row));
 }
 
 static const char *
-- 
2.31.0
