From e390b8e76f796e5b1d619cf36a0ed46103029e0b Mon Sep 17 00:00:00 2001
From: Mohammed Sadiq <sadiq@sadiqpk.org>
Date: Wed, 25 Sep 2019 21:27:13 +0530
Subject: [PATCH 3/8] notification: Hide lockscreen notification option on
 phones

---
 panels/notifications/cc-notifications-panel.c | 3 +++
 1 file changed, 3 insertions(+)

Index: gnome-control-center/panels/notifications/cc-notifications-panel.c
===================================================================
--- gnome-control-center.orig/panels/notifications/cc-notifications-panel.c	2021-07-29 04:01:34.853000000 +0000
+++ gnome-control-center/panels/notifications/cc-notifications-panel.c	2021-07-29 04:01:51.540000000 +0000
@@ -18,6 +18,7 @@
  */
 
 #include "config.h"
+#include "pureos.h"
 
 #include <string.h>
 #include <glib/gi18n-lib.h>
@@ -209,6 +210,9 @@
                             cc_panel_get_cancellable (CC_PANEL (panel)),
                             on_perm_store_ready,
                             panel);
+
+  if (pureos_get_is_phone ())
+    gtk_widget_hide (GTK_WIDGET (panel->lock_screen_row));
 }
 
 static const char *
