diff --git i/panels/info-overview/meson.build w/panels/info-overview/meson.build
index af5213bf3..899e77b79 100644
--- i/panels/info-overview/meson.build
+++ w/panels/info-overview/meson.build
@@ -69,7 +69,7 @@ executable(
 
 install_data(
   [ 'gnome-logo-text.svg', 'gnome-logo-text-dark.svg' ],
-  install_dir: join_paths(control_center_datadir, 'icons')
+  install_dir: join_paths(control_center_datadir, 'pixmaps')
 )
 
 subdir('icons')
