diff --git a/src/examples/media-session/media-session.c b/src/examples/media-session/media-session.c
index 9fa893a0..3d30911d 100644
--- a/src/examples/media-session/media-session.c
+++ b/src/examples/media-session/media-session.c
@@ -2019,7 +2019,7 @@ static void do_quit(void *data, int signal_number)
 				"default-profile,"	\
 				"default-routes,"	\
 				"restore-stream,"	\
-				"alsa-acp,"		\
+				"alsa-pcm,"		\
 				"alsa-seq,"		\
 				"v4l2,"			\
 				"bluez5,"		\
