From 9848a73b737ad46b5790ebf713a559d340c91b82 Mon Sep 17 00:00:00 2001
From: Samuel Dionne-Riel <samuel@dionne-riel.com>
Date: Sat, 3 Apr 2021 17:46:54 -0400
Subject: [PATCH] Don't crash when no language plugin is loaded

---
 src/lib/logic/wordengine.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/lib/logic/wordengine.cpp b/src/lib/logic/wordengine.cpp
index 16bf91e5..d04b603f 100644
--- a/src/lib/logic/wordengine.cpp
+++ b/src/lib/logic/wordengine.cpp
@@ -189,7 +189,7 @@ void WordEngine::setWordPredictionEnabled(bool enabled)
         enabled = false;
     }
 
-    if (d->languagePlugin->languageFeature()->alwaysShowSuggestions()) {
+    if (d->languagePlugin && d->languagePlugin->languageFeature()->alwaysShowSuggestions()) {
         // Override requested setting for languages that should always
         // display suggestions, such as Pinyin
         enabled = true;
