From ae75e44d8aca9d514c5298f9ee9a5a74bca30a8f Mon Sep 17 00:00:00 2001
From: Nbiba Bedis <bedisnbiba@gmail.com>
Date: Thu, 30 Sep 2021 21:53:55 +0100
Subject: [PATCH] container: don't panic if binder device already exists

---
 tools/helpers/drivers.py | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/tools/helpers/drivers.py b/tools/helpers/drivers.py
index 31005a5..25a5b62 100644
--- a/tools/helpers/drivers.py
+++ b/tools/helpers/drivers.py
@@ -58,7 +58,10 @@ def IOWR(_type, nr, size):
     for node in binder_dev_nodes:
         node_struct = struct.pack(
             '256sII', bytes(node, 'utf-8'), 0, 0)
-        fcntl.ioctl(binderctrlfd.fileno(), BINDER_CTL_ADD, node_struct)
+        try:
+            fcntl.ioctl(binderctrlfd.fileno(), BINDER_CTL_ADD, node_struct)
+        except FileExistsError:
+            pass
 
 def probeBinderDriver(args):
     binder_dev_nodes = []
