diff --text --unified --recursive f_scripts-0.4/scripts/f_audio f_scripts-0.4.new/scripts/f_audio
--- f_scripts-0.4/scripts/f_audio	2022-03-16 01:53:04.000000000 +0000
+++ f_scripts-0.4.new/scripts/f_audio	2022-06-26 04:51:19.216835863 +0000
@@ -1,6 +1,5 @@
 #!/usr/bin/env osh
-shopt -s strict:all
-DEP="alsa-utils"
+DEP="'alsa-utils'"
 DEC="Adjust audio output devices and volume"
 DOC="
   Allows you to adjust audio output devices and volume by using
@@ -41,15 +40,17 @@
   amixer set "AIF1 Slot 0 Digital DAC" unmute > /dev/null
   amixer set DAC "100%" > /dev/null
   amixer set DAC unmute > /dev/null
-  GLOBAL_SPEAKER="Line Out"
-  GLOBAL_HEADPHONE="Headphone"
-  GLOBAL_EARPIECE="Earpiece"
 }
 
 main() {
   local RUN OPTS
+
+  GLOBAL_SPEAKER="Line Out"
+  GLOBAL_HEADPHONE="Headphone"
+  GLOBAL_EARPIECE="Earpiece"
+
   env | grep -q "^$(basename "$0" | tr '[a-z]' '[A-Z]')=" || eval "$VAR"
-  eval "$(grep deviceinfo_codename /etc/deviceinfo | cut -d= -f2 | tr -d \"- | xargs -ID echo deviceD)"
+  eval "$(grep PinePhone /proc/device-tree/model | cut -d= -f2 | tr -d \"- | xargs -ID echo deviceD)"
   OPTS="speaker headphone earpiece volup voldown"
   [ -p /tmp/fbp.fifo ] && { echo -e "\b\f\r"; echo "$OPTS" | tr " " "\n"; } > /tmp/fbp.fifo
   while true; do
diff --text --unified --recursive f_scripts-0.4/scripts/f_files f_scripts-0.4.new/scripts/f_files
--- f_scripts-0.4/scripts/f_files	2022-03-16 01:53:04.000000000 +0000
+++ f_scripts-0.4.new/scripts/f_files	2022-06-26 04:53:23.867988884 +0000
@@ -1,6 +1,6 @@
 #!/usr/bin/env osh
 shopt -s strict:all
-DEP="vim mpv fbida-fbi fbida-fbgs"
+DEP="'vim' 'mpv' 'fbida'"
 DEC="Browse files on your filesystem"
 DOC="
   An interactive prompt-driven filebrowser. Enter the name of the
diff --text --unified --recursive f_scripts-0.4/scripts/f_game f_scripts-0.4.new/scripts/f_game
--- f_scripts-0.4/scripts/f_game	2022-03-16 01:53:04.000000000 +0000
+++ f_scripts-0.4.new/scripts/f_game	2022-06-26 04:53:23.879989781 +0000
@@ -1,6 +1,6 @@
 #!/usr/bin/env osh
 shopt -s strict:all
-DEP="moon-buggy"
+DEP="'moon-buggy'"
 DEC="Play classic console games"
 DOC="
   Launches classic console games with keybindings set in fbp.
diff --text --unified --recursive f_scripts-0.4/scripts/f_maps f_scripts-0.4.new/scripts/f_maps
--- f_scripts-0.4/scripts/f_maps	2022-03-16 01:53:04.000000000 +0000
+++ f_scripts-0.4.new/scripts/f_maps	2022-06-26 04:53:23.887990380 +0000
@@ -1,6 +1,6 @@
 #!/usr/bin/env osh
 shopt -s strict:all
-DEP="mepo tslib directfb mesa-gl kbd vis"
+DEP="'mepo' 'tslib' 'directfb' 'mesa' 'kbd' 'vis'"
 DEC="Browse OSM maps using mepo (via SDL directfb)."
 DOC="
   Launches mepo to browse OSM maps in SDL / directfb mode.  Uses chvt
diff --text --unified --recursive f_scripts-0.4/scripts/f_networks f_scripts-0.4.new/scripts/f_networks
--- f_scripts-0.4/scripts/f_networks	2022-03-16 01:53:04.000000000 +0000
+++ f_scripts-0.4.new/scripts/f_networks	2022-06-26 04:53:23.899991277 +0000
@@ -1,6 +1,6 @@
 #!/usr/bin/env osh
 shopt -s strict:all
-DEP="networkmanager"
+DEP="'networkmanager' 'ntp'"
 DEC="Allows managing gsm/wpa connections via underlying nmcli calls"
 DOC="
   Allows mangaging gsm/wpa connectiona via nmcli. Allows for creating,
@@ -46,7 +46,8 @@
       });
     ' | sudo tee "$TARGETFILE"
     sudo chmod -R +r "$(dirname "$TARGETFILE")"
-    sudo addgroup "$USER" plugdev
+    sudo groupadd plugdev
+    sudo gpasswd -a "$USER" plugdev
   }
 }
 
@@ -97,13 +98,13 @@
 }
 
 ensurenmrunning() {
-  rc-service networkmanager status | grep started &&
-  rc-service ntpd status | grep started || 
+  systemctl status NetworkManager | grep running &&
+  systemctl status ntpd | grep running || 
   {
-    sudo /etc/init.d/networkmanager start
-    sudo /etc/init.d/ntpd start
-    sudo rc-update add networkmanager boot
-    sudo rc-update add ntpd boot
+    sudo systemctl start NetworkManager
+    sudo systemctl start ntpd
+    sudo systemctl enable NetworkManager
+    sudo systemctl enable ntpd
   }
 }
 
diff --text --unified --recursive f_scripts-0.4/scripts/f_phone f_scripts-0.4.new/scripts/f_phone
--- f_scripts-0.4/scripts/f_phone	2022-03-16 01:53:04.000000000 +0000
+++ f_scripts-0.4.new/scripts/f_phone	2022-06-26 04:53:23.919992774 +0000
@@ -1,6 +1,6 @@
 #!/usr/bin/env osh
 shopt -s strict:all
-DEP="polkit modemmanager linuxconsoletools pinephone-call-audio pn"
+DEP="'polkit' 'modemmanager' 'linuxconsole' 'pinephone-call-audio' 'pn'"
 DEC="Manages phone calls / text messages using modemmanager"
 DOC="
   A single interactive prompt-driven script to manage all aspects of
@@ -73,7 +73,8 @@
       });
     ' | sudo tee "$TARGETPOLKITFILE"
     sudo chmod a+rx "$TARGETPOLKITFILE"
-    sudo addgroup "$USER" plugdev
+    sudo groupadd plugdev
+    sudo gpasswd -a "$USER" plugdev
     clear
     echo "polkit rule installed! you may need to reboot"
     promptok
@@ -95,9 +96,9 @@
 }
 
 ensuremmrunning() {
-  rc-service modemmanager status | grep started || {
-    sudo /etc/init.d/modemmanager start
-    sudo rc-update add modemmanager boot
+  systemctl status ModemManager | grep running || {
+    sudo systemctl start ModemManager
+    sudo systemctl enable ModemManager
   }
 }
 
@@ -196,7 +197,7 @@
 
 modemrst() {
   mmcli -L | grep /org/freedesktop/ModemManager1 ||
-    sudo /etc/init.d/modemmanager restart
+    sudo systemctl restart ModemManager
 
   while ! mmcli -L | grep /org/freedesktop/ModemManager1; do
     echo "Waiting for modem to come online"
@@ -424,7 +425,7 @@
   ensuremmrunning
   modemrst
   env | grep -q "^$(basename "$0" | tr '[a-z]' '[A-Z]')=" || eval "$VAR"
-  eval "$(grep deviceinfo_codename /etc/deviceinfo | cut -d= -f2 | tr -d \"- | xargs -ID echo deviceD)" || 
+  eval "$(grep PinePhone /proc/device-tree/model | cut -d= -f2 | tr -d \"- | xargs -ID echo deviceD)" || 
     errdie "Device unsupported - only pinephone supported currently"
 
   monitorboot $$ >&2 2>/dev/null &
diff --text --unified --recursive f_scripts-0.4/scripts/f_rss f_scripts-0.4.new/scripts/f_rss
--- f_scripts-0.4/scripts/f_rss	2022-03-16 01:53:04.000000000 +0000
+++ f_scripts-0.4.new/scripts/f_rss	2022-06-26 04:53:23.923993073 +0000
@@ -1,6 +1,6 @@
 #!/usr/bin/env osh
 shopt -s strict:all
-DEP="sfeed w3m"
+DEP="'sfeed' 'w3m'"
 DEC="Fetches data from RSS via sfeed and allows you to view via sfeed_curses"
 DOC="
   Allows you to fetch RSS feeds via sfeed and then view via sfeed_curses with
diff --text --unified --recursive f_scripts-0.4/scripts/f_theme f_scripts-0.4.new/scripts/f_theme
--- f_scripts-0.4/scripts/f_theme	2022-03-16 01:53:04.000000000 +0000
+++ f_scripts-0.4.new/scripts/f_theme	2022-06-26 04:53:23.935993970 +0000
@@ -1,6 +1,6 @@
 #!/usr/bin/env osh
 shopt -s strict:all
-DEP="terminus-font util-linux-misc"
+DEP="'terminus-font' 'util-linux'"
 DEC="Change the framebuffer theme (font & colors)"
 DOC="
   Allows customizing your framebuffer's theme by setting the font, colors,
@@ -12,21 +12,21 @@
   used for {sm,md,lg}{white,dark}.
 "
 VAR='
-  F_THEME_FONTSM="/usr/share/consolefonts/ter-114n.psf.gz"
-  F_THEME_FONTMD="/usr/share/consolefonts/ter-920n.psf.gz"
-  F_THEME_FONTLG="/usr/share/consolefonts/ter-v32n.psf.gz"
+  F_THEME_FONTSM="/usr/share/kbd/consolefonts/ter-114n.psf.gz"
+  F_THEME_FONTMD="/usr/share/kbd/consolefonts/ter-920n.psf.gz"
+  F_THEME_FONTLG="/usr/share/kbd/consolefonts/ter-v32n.psf.gz"
 '
 
 fontchange() {
   local FONTS PICKED
 
-  FONTS="$(ls -1 /usr/share/consolefonts/ | sed 's/.psf.gz//g' )"
+  FONTS="$(ls -1 /usr/share/kbd/consolefonts/ | sed 's/.psf.gz//g' )"
   [ -p /tmp/fbp.fifo ] && echo -e "\b\fcancel\n$FONTS" > /tmp/fbp.fifo
   echo "$FONTS"
   echo "Pick a font:"
   read -r PICKED
   [ "$PICKED" = cancel ] && return
-  setfont /usr/share/consolefonts/$PICKED.psf.gz
+  setfont /usr/share/kbd/consolefonts/$PICKED.psf.gz
   [ -p /tmp/fbp.fifo ] && echo -e "\v" > /tmp/fbp.fifo
 }
 
diff --text --unified --recursive f_scripts-0.4/scripts/f_timer f_scripts-0.4.new/scripts/f_timer
--- f_scripts-0.4/scripts/f_timer	2022-03-16 01:53:04.000000000 +0000
+++ f_scripts-0.4.new/scripts/f_timer	2022-06-26 04:53:23.943994569 +0000
@@ -1,6 +1,6 @@
 #!/usr/bin/env osh
 shopt -s strict:all
-DEP="linuxconsoletools util-linux-misc"
+DEP="'linuxconsole' 'util-linux' 'bc'"
 DEC="Set a countdown timer to be alerted"
 DOC="
   Set a countdown timer after which expiry you'll be alerted.
diff --text --unified --recursive f_scripts-0.4/scripts/f_web f_scripts-0.4.new/scripts/f_web
--- f_scripts-0.4/scripts/f_web	2022-03-16 01:53:04.000000000 +0000
+++ f_scripts-0.4.new/scripts/f_web	2022-06-26 04:53:23.951995167 +0000
@@ -1,6 +1,6 @@
 #!/usr/bin/env osh
 shopt -s strict:all
-DEP="w3m netsurf-framebuffer tslib"
+DEP="'w3m' 'netsurf-fb' 'tslib'"
 DEC="Select a website and browse via w3m or netsurf"
 DOC="
   Browse the web via w3m (text) or netsurf (graphical) setting fbp hotkeys
diff --text --unified --recursive f_scripts-0.4/scripts/f_youtube f_scripts-0.4.new/scripts/f_youtube
--- f_scripts-0.4/scripts/f_youtube	2022-03-16 01:53:04.000000000 +0000
+++ f_scripts-0.4.new/scripts/f_youtube	2022-06-26 04:53:23.955995467 +0000
@@ -1,6 +1,6 @@
 #!/usr/bin/env osh
 shopt -s strict:all
-DEP="codemadness-frontends mpv youtube-dl"
+DEP="'codemadness-frontends' 'mpv' 'youtube-dl'"
 DEC="Listen or view youtube videos via mpv"
 DOC="
   f_youtube allows you to search and playback either audio or video from
@@ -73,7 +73,7 @@
   local AV QUERY
 
   env | grep -q "^$(basename "$0" | tr '[a-z]' '[A-Z]')=" || eval "$VAR"
-  eval "$(grep deviceinfo_codename /etc/deviceinfo | cut -d= -f2 | tr -d \"- | xargs -ID echo deviceD)"
+  eval "$(grep PinePhone /proc/device-tree/model | cut -d= -f2 | tr -d \"- | xargs -ID echo deviceD)"
 
   OPTS="audio video"
   [ -p /tmp/fbp.fifo ] && { echo -e "\b\f\r"; echo "$OPTS" | tr " " "\n"; } > /tmp/fbp.fifo
