diff --git a/arch/arm64/boot/dts/allwinner/sun50i-a64-pinephone.dtsi b/arch/arm64/boot/dts/allwinner/sun50i-a64-pinephone.dtsi
index 086b5ebfa512..f929af5c34bb 100644
--- a/arch/arm64/boot/dts/allwinner/sun50i-a64-pinephone.dtsi
+++ b/arch/arm64/boot/dts/allwinner/sun50i-a64-pinephone.dtsi
@@ -592,11 +592,7 @@ &pio {
 	vcc-pb-supply = <&reg_dcdc1>;
 	vcc-pc-supply = <&reg_dcdc1>;
 	vcc-pd-supply = <&reg_dcdc1>;
-	/* pinctrl would enable this even if no camera is powered,
-	 * which is wrong/not necessary
-	 *
-	 * vcc-pe-supply = <&reg_aldo1>; (also used by pogo pins i2c)
-	 */
+	vcc-pe-supply = <&reg_aldo1>; /* (also used by pogo pins i2c) */
 	vcc-pf-supply = <&reg_dcdc1>;
 	vcc-pg-supply = <&reg_dldo4>;
 	vcc-ph-supply = <&reg_dcdc1>;
