From f771fd05732e9709d9109525a413e5bc381958d8 Mon Sep 17 00:00:00 2001
From: Evangelos Ribeiro Tzaras <evangelos.tzaras@puri.sm>
Date: Sat, 10 Jul 2021 04:29:50 +0200
Subject: [PATCH 1/7] meson: Change homepage to GNOMEs gitlab

---
 meson.build | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/meson.build b/meson.build
index 67dcecff..a9122a47 100644
--- a/meson.build
+++ b/meson.build
@@ -33,7 +33,7 @@ project(
 )
 
 calls_id = 'sm.puri.Calls'
-calls_homepage = 'https://source.puri.sm/Librem5/calls'
+calls_homepage = 'https://gitlab.gnome.org/GNOME/calls'
 calls_name = meson.project_name()
 calls_version = meson.project_version()
 
-- 
GitLab


From bd9dd8475f5393a6760b8f27d4c90a6aac750596 Mon Sep 17 00:00:00 2001
From: Evangelos Ribeiro Tzaras <evangelos.tzaras@puri.sm>
Date: Sat, 10 Jul 2021 04:42:50 +0200
Subject: [PATCH 2/7] docs: Change homepage to GNOMEs gitlab

---
 doc/xml/meson.build | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/doc/xml/meson.build b/doc/xml/meson.build
index 55537870..7b2ce251 100644
--- a/doc/xml/meson.build
+++ b/doc/xml/meson.build
@@ -1,9 +1,9 @@
 ent_conf = configuration_data()
 ent_conf.set('PACKAGE', 'Calls')
-ent_conf.set('PACKAGE_BUGREPORT', 'https://source.puri.sm/Librem5/calls/issues')
+ent_conf.set('PACKAGE_BUGREPORT', 'https://gitlab.gnome.org/GNOME/calls/issues')
 ent_conf.set('PACKAGE_NAME', 'Calls')
 ent_conf.set('PACKAGE_STRING', 'calls')
 ent_conf.set('PACKAGE_TARNAME', 'calls-' + meson.project_version())
-ent_conf.set('PACKAGE_URL', 'https://source.puri.sm/Librem5/calls')
+ent_conf.set('PACKAGE_URL', 'https://gitlab.gnome.org/GNOME/calls')
 ent_conf.set('PACKAGE_VERSION', meson.project_version())
 configure_file(input: 'gtkdocentities.ent.in', output: 'gtkdocentities.ent', configuration: ent_conf)
-- 
GitLab


From 6e371d2f3fae3e863b20438b040c26a586b5af46 Mon Sep 17 00:00:00 2001
From: Evangelos Ribeiro Tzaras <evangelos.tzaras@puri.sm>
Date: Sat, 10 Jul 2021 04:15:22 +0200
Subject: [PATCH 3/7] resources: Use /org/gnome prefix and adapt to changes

In preparation to moving to GNOME all the IDs, DBus object paths, etc should
change.
---
 src/calls-application.c          | 2 +-
 src/calls-call-display.c         | 2 +-
 src/calls-call-record-row.c      | 2 +-
 src/calls-call-selector-item.c   | 2 +-
 src/calls-call-window.c          | 2 +-
 src/calls-contacts-box.c         | 2 +-
 src/calls-contacts-row.c         | 2 +-
 src/calls-encryption-indicator.c | 2 +-
 src/calls-history-box.c          | 2 +-
 src/calls-in-app-notification.c  | 2 +-
 src/calls-main-window.c          | 2 +-
 src/calls-new-call-box.c         | 2 +-
 src/calls.gresources.xml         | 6 +++---
 13 files changed, 15 insertions(+), 15 deletions(-)

diff --git a/src/calls-application.c b/src/calls-application.c
index 7042f0c4..54ad8617 100644
--- a/src/calls-application.c
+++ b/src/calls-application.c
@@ -409,7 +409,7 @@ startup (GApplication *application)
   manager_state_changed_cb (application);
 
   provider = gtk_css_provider_new ();
-  gtk_css_provider_load_from_resource (provider, "/sm/puri/calls/style.css");
+  gtk_css_provider_load_from_resource (provider, "/org/gnome/Calls/style.css");
   gtk_style_context_add_provider_for_screen (gdk_screen_get_default (),
                                              GTK_STYLE_PROVIDER (provider),
                                              GTK_STYLE_PROVIDER_PRIORITY_APPLICATION);
diff --git a/src/calls-call-display.c b/src/calls-call-display.c
index bf989582..466fbe28 100644
--- a/src/calls-call-display.c
+++ b/src/calls-call-display.c
@@ -485,7 +485,7 @@ calls_call_display_class_init (CallsCallDisplayClass *klass)
   g_object_class_install_properties (object_class, PROP_LAST_PROP, props);
 
 
-  gtk_widget_class_set_template_from_resource (widget_class, "/sm/puri/calls/ui/call-display.ui");
+  gtk_widget_class_set_template_from_resource (widget_class, "/org/gnome/Calls/ui/call-display.ui");
   gtk_widget_class_bind_template_child (widget_class, CallsCallDisplay, incoming_phone_call);
   gtk_widget_class_bind_template_child (widget_class, CallsCallDisplay, primary_contact_info);
   gtk_widget_class_bind_template_child (widget_class, CallsCallDisplay, secondary_contact_info);
diff --git a/src/calls-call-record-row.c b/src/calls-call-record-row.c
index ebad58d5..155e7840 100644
--- a/src/calls-call-record-row.c
+++ b/src/calls-call-record-row.c
@@ -520,7 +520,7 @@ calls_call_record_row_class_init (CallsCallRecordRowClass *klass)
   g_object_class_install_properties (object_class, PROP_LAST_PROP, props);
 
 
-  gtk_widget_class_set_template_from_resource (widget_class, "/sm/puri/calls/ui/call-record-row.ui");
+  gtk_widget_class_set_template_from_resource (widget_class, "/org/gnome/Calls/ui/call-record-row.ui");
   gtk_widget_class_bind_template_child (widget_class, CallsCallRecordRow, avatar);
   gtk_widget_class_bind_template_child (widget_class, CallsCallRecordRow, type);
   gtk_widget_class_bind_template_child (widget_class, CallsCallRecordRow, target);
diff --git a/src/calls-call-selector-item.c b/src/calls-call-selector-item.c
index f5410a78..bc928d2f 100644
--- a/src/calls-call-selector-item.c
+++ b/src/calls-call-selector-item.c
@@ -179,7 +179,7 @@ calls_call_selector_item_class_init (CallsCallSelectorItemClass *klass)
   g_object_class_install_properties (object_class, PROP_LAST_PROP, props);
 
 
-  gtk_widget_class_set_template_from_resource (widget_class, "/sm/puri/calls/ui/call-selector-item.ui");
+  gtk_widget_class_set_template_from_resource (widget_class, "/org/gnome/Calls/ui/call-selector-item.ui");
   gtk_widget_class_bind_template_child (widget_class, CallsCallSelectorItem, main_box);
   gtk_widget_class_bind_template_child (widget_class, CallsCallSelectorItem, name);
   gtk_widget_class_bind_template_child (widget_class, CallsCallSelectorItem, status);
diff --git a/src/calls-call-window.c b/src/calls-call-window.c
index 1f49cf2d..e9c63139 100644
--- a/src/calls-call-window.c
+++ b/src/calls-call-window.c
@@ -577,7 +577,7 @@ calls_call_window_class_init (CallsCallWindowClass *klass)
   object_class->notify = notify;
 #endif // CALLS_WAYLAND
 
-  gtk_widget_class_set_template_from_resource (widget_class, "/sm/puri/calls/ui/call-window.ui");
+  gtk_widget_class_set_template_from_resource (widget_class, "/org/gnome/Calls/ui/call-window.ui");
   gtk_widget_class_bind_template_child (widget_class, CallsCallWindow, in_app_notification);
   gtk_widget_class_bind_template_child (widget_class, CallsCallWindow, main_stack);
   gtk_widget_class_bind_template_child (widget_class, CallsCallWindow, header_bar_stack);
diff --git a/src/calls-contacts-box.c b/src/calls-contacts-box.c
index f346d91e..e6429cae 100644
--- a/src/calls-contacts-box.c
+++ b/src/calls-contacts-box.c
@@ -145,7 +145,7 @@ calls_contacts_box_class_init (CallsContactsBoxClass *klass)
 {
   GtkWidgetClass *widget_class = GTK_WIDGET_CLASS (klass);
 
-  gtk_widget_class_set_template_from_resource (widget_class, "/sm/puri/calls/ui/contacts-box.ui");
+  gtk_widget_class_set_template_from_resource (widget_class, "/org/gnome/Calls/ui/contacts-box.ui");
   gtk_widget_class_bind_template_child (widget_class, CallsContactsBox, contacts_listbox);
   gtk_widget_class_bind_template_child (widget_class, CallsContactsBox, contacts_frame);
   gtk_widget_class_bind_template_child (widget_class, CallsContactsBox, search_entry);
diff --git a/src/calls-contacts-row.c b/src/calls-contacts-row.c
index e2c4b00c..6a5b54ad 100644
--- a/src/calls-contacts-row.c
+++ b/src/calls-contacts-row.c
@@ -127,7 +127,7 @@ calls_contacts_row_class_init (CallsContactsRowClass *klass)
   object_class->dispose = calls_contacts_row_dispose;
 
   gtk_widget_class_set_template_from_resource (widget_class,
-                                               "/sm/puri/calls/"
+                                               "/org/gnome/Calls/"
                                                "ui/contacts-row.ui");
   gtk_widget_class_bind_template_child (widget_class, CallsContactsRow, avatar);
   gtk_widget_class_bind_template_child (widget_class, CallsContactsRow, title);
diff --git a/src/calls-encryption-indicator.c b/src/calls-encryption-indicator.c
index 68676b6b..b39267f2 100644
--- a/src/calls-encryption-indicator.c
+++ b/src/calls-encryption-indicator.c
@@ -131,7 +131,7 @@ calls_encryption_indicator_class_init (CallsEncryptionIndicatorClass *klass)
                           G_PARAM_READWRITE);
 
   g_object_class_install_properties (object_class, PROP_LAST_PROP, props);
-  gtk_widget_class_set_template_from_resource (widget_class, "/sm/puri/calls/ui/encryption-indicator.ui");
+  gtk_widget_class_set_template_from_resource (widget_class, "/org/gnome/Calls/ui/encryption-indicator.ui");
   gtk_widget_class_bind_template_child (widget_class, CallsEncryptionIndicator, is_not_encrypted);
   gtk_widget_class_bind_template_child (widget_class, CallsEncryptionIndicator, is_encrypted);
 }
diff --git a/src/calls-history-box.c b/src/calls-history-box.c
index e9dc900c..48602a61 100644
--- a/src/calls-history-box.c
+++ b/src/calls-history-box.c
@@ -226,7 +226,7 @@ calls_history_box_class_init (CallsHistoryBoxClass *klass)
   g_object_class_install_properties (object_class, PROP_LAST_PROP, props);
 
 
-  gtk_widget_class_set_template_from_resource (widget_class, "/sm/puri/calls/ui/history-box.ui");
+  gtk_widget_class_set_template_from_resource (widget_class, "/org/gnome/Calls/ui/history-box.ui");
   gtk_widget_class_bind_template_child (widget_class, CallsHistoryBox, history);
 }
 
diff --git a/src/calls-in-app-notification.c b/src/calls-in-app-notification.c
index 14994eb0..018ccf02 100644
--- a/src/calls-in-app-notification.c
+++ b/src/calls-in-app-notification.c
@@ -128,7 +128,7 @@ calls_in_app_notification_class_init (CallsInAppNotificationClass *klass)
 
   g_object_class_install_properties (object_class, PROP_LAST_PROP, props);
 
-  gtk_widget_class_set_template_from_resource (widget_class, "/sm/puri/calls/ui/in-app-notification.ui");
+  gtk_widget_class_set_template_from_resource (widget_class, "/org/gnome/Calls/ui/in-app-notification.ui");
   gtk_widget_class_bind_template_child (widget_class, CallsInAppNotification, label);
   gtk_widget_class_bind_template_callback (widget_class, calls_in_app_notification_hide);
 }
diff --git a/src/calls-main-window.c b/src/calls-main-window.c
index fb2da480..57c3fd35 100644
--- a/src/calls-main-window.c
+++ b/src/calls-main-window.c
@@ -466,7 +466,7 @@ calls_main_window_class_init (CallsMainWindowClass *klass)
 
   widget_class->size_allocate = size_allocate;
 
-  gtk_widget_class_set_template_from_resource (widget_class, "/sm/puri/calls/ui/main-window.ui");
+  gtk_widget_class_set_template_from_resource (widget_class, "/org/gnome/Calls/ui/main-window.ui");
   gtk_widget_class_bind_template_child (widget_class, CallsMainWindow, in_app_notification);
   gtk_widget_class_bind_template_child (widget_class, CallsMainWindow, title_switcher);
   gtk_widget_class_bind_template_child (widget_class, CallsMainWindow, main_stack);
diff --git a/src/calls-new-call-box.c b/src/calls-new-call-box.c
index 91c7691e..c5fa1206 100644
--- a/src/calls-new-call-box.c
+++ b/src/calls-new-call-box.c
@@ -270,7 +270,7 @@ calls_new_call_box_class_init (CallsNewCallBoxClass *klass)
 
   object_class->dispose = dispose;
 
-  gtk_widget_class_set_template_from_resource (widget_class, "/sm/puri/calls/ui/new-call-box.ui");
+  gtk_widget_class_set_template_from_resource (widget_class, "/org/gnome/Calls/ui/new-call-box.ui");
   gtk_widget_class_bind_template_child (widget_class, CallsNewCallBox, origin_list_box);
   gtk_widget_class_bind_template_child (widget_class, CallsNewCallBox, origin_list);
   gtk_widget_class_bind_template_child (widget_class, CallsNewCallBox, backspace);
diff --git a/src/calls.gresources.xml b/src/calls.gresources.xml
index e413bc0c..9a4e4ca1 100644
--- a/src/calls.gresources.xml
+++ b/src/calls.gresources.xml
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <gresources>
-  <gresource prefix="/sm/puri/calls/ui">
+  <gresource prefix="/org/gnome/Calls/ui">
     <file preprocess="xml-stripblanks">main-window.ui</file>
     <file preprocess="xml-stripblanks">call-display.ui</file>
     <file preprocess="xml-stripblanks">call-selector-item.ui</file>
@@ -15,10 +15,10 @@
     <file preprocess="xml-stripblanks">contacts-row.ui</file>
     <file preprocess="xml-stripblanks">contacts-box.ui</file>
   </gresource>
-  <gresource prefix="/sm/puri/calls/">
+  <gresource prefix="/org/gnome/Calls/">
     <file>style.css</file>
   </gresource>
-  <gresource prefix="/sm/puri/Calls/icons/">
+  <gresource prefix="/org/gnome/Calls/icons/">
     <file>new-call-symbolic.svg</file>
     <file>call-arrow-incoming-symbolic.svg</file>
     <file>call-arrow-incoming-missed-symbolic.svg</file>
-- 
GitLab


From b74601f7c42705b6ce8f3fb3af8a92a959c84090 Mon Sep 17 00:00:00 2001
From: Evangelos Ribeiro Tzaras <evangelos.tzaras@puri.sm>
Date: Sat, 10 Jul 2021 04:19:53 +0200
Subject: [PATCH 4/7] schema: Use org.gnome prefix and adapt to changes

In preparation to moving to GNOME all the IDs, DBus object paths, etc should
change.
---
 data/meson.build                                                | 2 +-
 data/{sm.puri.Calls.gschema.xml => org.gnome.Calls.gschema.xml} | 2 +-
 src/calls-settings.c                                            | 2 +-
 3 files changed, 3 insertions(+), 3 deletions(-)
 rename data/{sm.puri.Calls.gschema.xml => org.gnome.Calls.gschema.xml} (93%)

diff --git a/data/meson.build b/data/meson.build
index 6f593cb3..5cdfeab5 100644
--- a/data/meson.build
+++ b/data/meson.build
@@ -68,7 +68,7 @@ install_data(
   )
 )
 
-schema_src = 'sm.puri.Calls.gschema.xml'
+schema_src = 'org.gnome.Calls.gschema.xml'
 compiled = gnome.compile_schemas(build_by_default: true,
                                  depend_files: files(schema_src))
 
diff --git a/data/sm.puri.Calls.gschema.xml b/data/org.gnome.Calls.gschema.xml
similarity index 93%
rename from data/sm.puri.Calls.gschema.xml
rename to data/org.gnome.Calls.gschema.xml
index 282e449a..eee54406 100644
--- a/data/sm.puri.Calls.gschema.xml
+++ b/data/org.gnome.Calls.gschema.xml
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <schemalist gettext-domain="purism-calls">
-  <schema id="sm.puri.Calls" path="/sm/puri/Calls/">
+  <schema id="org.gnome.Calls" path="/org/gnome/Calls/">
 
     <key name="auto-use-default-origins" type="b">
       <default>true</default>
diff --git a/src/calls-settings.c b/src/calls-settings.c
index c33d7668..8aa3dafb 100644
--- a/src/calls-settings.c
+++ b/src/calls-settings.c
@@ -169,7 +169,7 @@ calls_settings_class_init (CallsSettingsClass *klass)
 static void
 calls_settings_init (CallsSettings *self)
 {
-  self->settings = g_settings_new ("sm.puri.Calls");
+  self->settings = g_settings_new ("org.gnome.Calls");
 }
 
 /**
-- 
GitLab


From 69b919d7ab58399fd3c41db6763bfa371fa8c034 Mon Sep 17 00:00:00 2001
From: Evangelos Ribeiro Tzaras <evangelos.tzaras@puri.sm>
Date: Sat, 10 Jul 2021 04:27:42 +0200
Subject: [PATCH 5/7] DBus: Use org.gnome prefix and adapt to changes

In preparation to moving to GNOME all the IDs, DBus object paths, etc should
change.
---
 doc/calls-docs.xml                                            | 2 +-
 src/calls-dbus-manager.c                                      | 4 ++--
 src/dbus/meson.build                                          | 4 ++--
 src/dbus/{sm.puri.Calls.Call.xml => org.gnome.Calls.Call.xml} | 2 +-
 4 files changed, 6 insertions(+), 6 deletions(-)
 rename src/dbus/{sm.puri.Calls.Call.xml => org.gnome.Calls.Call.xml} (97%)

diff --git a/doc/calls-docs.xml b/doc/calls-docs.xml
index f2e6b571..2585ff9b 100644
--- a/doc/calls-docs.xml
+++ b/doc/calls-docs.xml
@@ -37,7 +37,7 @@
 
   <chapter id="dbus">
     <title>DBus interfaces</title>
-    <xi:include href="../src/dbus/calls-sm.puri.Calls.Call.xml"/>
+    <xi:include href="../src/dbus/calls-org.gnome.Calls.Call.xml"/>
   </chapter>
 
 </book>
diff --git a/src/calls-dbus-manager.c b/src/calls-dbus-manager.c
index 9bced163..5ea9b3c9 100644
--- a/src/calls-dbus-manager.c
+++ b/src/calls-dbus-manager.c
@@ -22,8 +22,8 @@
  * The #CallsCallManager is responsible for exposing the
  * call objects on the session bus using #GDBusObjectManager.
  *
- * Call objects are exported like /sm/puri/Calls/ as
- * /sm/puri/Calls/Call/1, /sm/puri/Calls/Call/2, … using
+ * Call objects are exported like /org/gnome/Calls/ as
+ * /org/gnome/Calls/Call/1, /org/gnome/Calls/Call/2, … using
  * org.freedesktop.DBus.ObjectManager.
  */
 
diff --git a/src/dbus/meson.build b/src/dbus/meson.build
index 64839f31..d87de30d 100644
--- a/src/dbus/meson.build
+++ b/src/dbus/meson.build
@@ -2,8 +2,8 @@ generated_dbus_sources = []
 
 # DBus server protocols
 generated_dbus_sources += gnome.gdbus_codegen('calls-call-dbus',
-                'sm.puri.Calls.Call.xml',
+                'org.gnome.Calls.Call.xml',
                 docbook: 'calls',
-                interface_prefix: 'sm.puri',
+                interface_prefix: 'org.gnome',
                 object_manager: true,
                 namespace: 'CallsDBus')
diff --git a/src/dbus/sm.puri.Calls.Call.xml b/src/dbus/org.gnome.Calls.Call.xml
similarity index 97%
rename from src/dbus/sm.puri.Calls.Call.xml
rename to src/dbus/org.gnome.Calls.Call.xml
index 78ecafb5..f8ca6fdd 100644
--- a/src/dbus/sm.puri.Calls.Call.xml
+++ b/src/dbus/org.gnome.Calls.Call.xml
@@ -21,7 +21,7 @@
 
 <node>
 
-  <interface name="sm.puri.Calls.Call">
+  <interface name="org.gnome.Calls.Call">
     <method name="Accept"/>
     <method name="Hangup"/>
     <property name="Inbound" type="b" access="read"/>
-- 
GitLab


From 7224998a1cb463b79c23c47f8e73b6da7380ca35 Mon Sep 17 00:00:00 2001
From: Evangelos Ribeiro Tzaras <evangelos.tzaras@puri.sm>
Date: Sat, 10 Jul 2021 04:34:23 +0200
Subject: [PATCH 6/7] project: Use org.gnome prefix in app id and adapt to
 changes

In preparation to moving to GNOME all the IDs, DBus object paths, etc should
change.

Closes #303
---
 data/meson.build                              | 20 +++++++++----------
 ...p.in => org.gnome.Calls-daemon.desktop.in} |  2 +-
 ...bolic.svg => org.gnome.Calls-symbolic.svg} |  0
 ....desktop.in => org.gnome.Calls.desktop.in} |  2 +-
 ...ainfo.xml => org.gnome.Calls.metainfo.xml} |  2 +-
 ...{sm.puri.Calls.svg => org.gnome.Calls.svg} |  2 +-
 meson.build                                   |  2 +-
 sm.puri.Calls.json => org.gnome.Calls.json    |  2 +-
 po/POTFILES.in                                |  4 ++--
 9 files changed, 17 insertions(+), 18 deletions(-)
 rename data/{sm.puri.Calls-daemon.desktop.in => org.gnome.Calls-daemon.desktop.in} (96%)
 rename data/{sm.puri.Calls-symbolic.svg => org.gnome.Calls-symbolic.svg} (100%)
 rename data/{sm.puri.Calls.desktop.in => org.gnome.Calls.desktop.in} (96%)
 rename data/{sm.puri.Calls.metainfo.xml => org.gnome.Calls.metainfo.xml} (99%)
 rename data/{sm.puri.Calls.svg => org.gnome.Calls.svg} (99%)
 rename sm.puri.Calls.json => org.gnome.Calls.json (99%)

diff --git a/data/meson.build b/data/meson.build
index 5cdfeab5..9aef3cfd 100644
--- a/data/meson.build
+++ b/data/meson.build
@@ -3,8 +3,8 @@ sysconfdir = get_option('sysconfdir')
 
 # Desktop file
 desktop_file = i18n.merge_file(
-  input : 'sm.puri.Calls.desktop.in',
-  output : 'sm.puri.Calls.desktop',
+  input : 'org.gnome.Calls.desktop.in',
+  output : 'org.gnome.Calls.desktop',
   type : 'desktop',
   po_dir : '../po',
   install : true,
@@ -12,8 +12,8 @@ desktop_file = i18n.merge_file(
 )
 
 desktop_file = i18n.merge_file(
-  input : 'sm.puri.Calls-daemon.desktop.in',
-  output : 'sm.puri.Calls-daemon.desktop',
+  input : 'org.gnome.Calls-daemon.desktop.in',
+  output : 'org.gnome.Calls-daemon.desktop',
   type : 'desktop',
   po_dir : '../po',
   install : true,
@@ -24,16 +24,16 @@ desktop_utils = find_program('desktop-file-validate', required: false)
 if desktop_utils.found()
   test('Validate desktop file', desktop_utils,
        args: [join_paths(meson.current_build_dir(),
-                         'sm.puri.Calls.desktop')
+                         'org.gnome.Calls.desktop')
              ])
   test('Validate daemon desktop file', desktop_utils,
        args: [join_paths(meson.current_build_dir(),
-                         'sm.puri.Calls-daemon.desktop')
+                         'org.gnome.Calls-daemon.desktop')
              ])
 endif
 
 # Metainfo file
-install_data('sm.puri.Calls.metainfo.xml',
+install_data('org.gnome.Calls.metainfo.xml',
     install_dir: join_paths(datadir, 'metainfo'),
 )
 
@@ -42,13 +42,13 @@ if appstream_util.found()
   test('Validate appstream file', appstream_util,
        args: ['validate-relax',
               join_paths(meson.current_source_dir(),
-                         'sm.puri.Calls.metainfo.xml')
+                         'org.gnome.Calls.metainfo.xml')
              ])
 endif
 
 # Icons
 install_data(
-  'sm.puri.Calls.svg',
+  'org.gnome.Calls.svg',
   install_dir: join_paths(
     datadir,
     'icons',
@@ -58,7 +58,7 @@ install_data(
   )
 )
 install_data(
-  'sm.puri.Calls-symbolic.svg',
+  'org.gnome.Calls-symbolic.svg',
   install_dir: join_paths(
     datadir,
     'icons',
diff --git a/data/sm.puri.Calls-daemon.desktop.in b/data/org.gnome.Calls-daemon.desktop.in
similarity index 96%
rename from data/sm.puri.Calls-daemon.desktop.in
rename to data/org.gnome.Calls-daemon.desktop.in
index 5d0d9c84..78dd32e1 100644
--- a/data/sm.puri.Calls-daemon.desktop.in
+++ b/data/org.gnome.Calls-daemon.desktop.in
@@ -5,7 +5,7 @@ Comment=A phone dialer and call handler (daemon mode)
 # Translators: These are desktop search terms. Do not translate semicolons, end line with a semicolon.
 Keywords=Telephone;Call;Phone;Dial;Dialer;PSTN;
 # Translators: Do NOT translate or transliterate this text (this is an icon file name)!
-Icon=sm.puri.Calls
+Icon=org.gnome.Calls
 TryExec=gnome-calls
 Exec=gnome-calls --daemon
 Type=Application
diff --git a/data/sm.puri.Calls-symbolic.svg b/data/org.gnome.Calls-symbolic.svg
similarity index 100%
rename from data/sm.puri.Calls-symbolic.svg
rename to data/org.gnome.Calls-symbolic.svg
diff --git a/data/sm.puri.Calls.desktop.in b/data/org.gnome.Calls.desktop.in
similarity index 96%
rename from data/sm.puri.Calls.desktop.in
rename to data/org.gnome.Calls.desktop.in
index 463e7675..f84ae136 100644
--- a/data/sm.puri.Calls.desktop.in
+++ b/data/org.gnome.Calls.desktop.in
@@ -5,7 +5,7 @@ Comment=A phone dialer and call handler
 # Translators: These are desktop search terms. Do not translate semicolons, end line with a semicolon.
 Keywords=Telephone;Call;Phone;Dial;Dialer;PSTN;
 # Translators: Do NOT translate or transliterate this text (this is an icon file name)!
-Icon=sm.puri.Calls
+Icon=org.gnome.Calls
 TryExec=gnome-calls
 Exec=gnome-calls %u
 Type=Application
diff --git a/data/sm.puri.Calls.metainfo.xml b/data/org.gnome.Calls.metainfo.xml
similarity index 99%
rename from data/sm.puri.Calls.metainfo.xml
rename to data/org.gnome.Calls.metainfo.xml
index a19f5881..8524971d 100644
--- a/data/sm.puri.Calls.metainfo.xml
+++ b/data/org.gnome.Calls.metainfo.xml
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <component type="desktop-application">
-    <id>sm.puri.Calls</id>
+    <id>org.gnome.Calls</id>
     <metadata_license>CC0-1.0</metadata_license>
     <project_license>GPL-3.0+</project_license>
     <name>Calls</name>
diff --git a/data/sm.puri.Calls.svg b/data/org.gnome.Calls.svg
similarity index 99%
rename from data/sm.puri.Calls.svg
rename to data/org.gnome.Calls.svg
index b44fbe4a..a2091435 100644
--- a/data/sm.puri.Calls.svg
+++ b/data/org.gnome.Calls.svg
@@ -18,7 +18,7 @@
    id="svg11300"
    sodipodi:version="0.32"
    inkscape:version="0.92.3 (2405546, 2018-03-11)"
-   sodipodi:docname="sm.puri.Calls.svg"
+   sodipodi:docname="org.gnome.Calls.svg"
    inkscape:output_extension="org.inkscape.output.svg.inkscape"
    version="1.0"
    style="display:inline;enable-background:new"
diff --git a/meson.build b/meson.build
index a9122a47..fa9bc61b 100644
--- a/meson.build
+++ b/meson.build
@@ -32,7 +32,7 @@ project(
   ]
 )
 
-calls_id = 'sm.puri.Calls'
+calls_id = 'org.gnome.Calls'
 calls_homepage = 'https://gitlab.gnome.org/GNOME/calls'
 calls_name = meson.project_name()
 calls_version = meson.project_version()
diff --git a/sm.puri.Calls.json b/org.gnome.Calls.json
similarity index 99%
rename from sm.puri.Calls.json
rename to org.gnome.Calls.json
index c7d4ef19..4eb4e7d4 100644
--- a/sm.puri.Calls.json
+++ b/org.gnome.Calls.json
@@ -1,5 +1,5 @@
 {
-    "app-id" : "sm.puri.Calls",
+    "app-id" : "org.gnome.Calls",
     "runtime" : "org.gnome.Platform",
     "runtime-version" : "master",
     "sdk" : "org.gnome.Sdk",
diff --git a/po/POTFILES.in b/po/POTFILES.in
index 595a7f31..d606a2a7 100644
--- a/po/POTFILES.in
+++ b/po/POTFILES.in
@@ -1,5 +1,5 @@
-data/sm.puri.Calls.desktop.in
-data/sm.puri.Calls-daemon.desktop.in
+data/org.gnome.Calls.desktop.in
+data/org.gnome.Calls-daemon.desktop.in
 src/calls-application.c
 src/calls-best-match.c
 src/calls-call.c
-- 
GitLab


From 4bfcf99155c3e6eaafc8936f1a06d9ad4825800d Mon Sep 17 00:00:00 2001
From: Evangelos Ribeiro Tzaras <evangelos.tzaras@puri.sm>
Date: Sat, 10 Jul 2021 12:41:17 +0200
Subject: [PATCH 7/7] debian: Remove old desktop files on upgrade

---
 debian/gnome-calls.maintscript | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/debian/gnome-calls.maintscript b/debian/gnome-calls.maintscript
index e8ac5101..04af8e9b 100644
--- a/debian/gnome-calls.maintscript
+++ b/debian/gnome-calls.maintscript
@@ -1 +1,2 @@
-mv_conffile /etc/xdg/autostart/sm.puri.Calls.desktop /etc/xdg/autostart/sm.puri.Calls-daemon.desktop 0.2.0~
+rm_conffile /etc/xdg/autostart/sm.puri.Calls.desktop 0.3.5~
+rm_conffile /etc/xdg/autostart/sm.puri.Calls-daemon.desktop 0.3.5~
-- 
GitLab

