diff --git a/src/fbd-feedback-manager.c b/src/fbd-feedback-manager.c
index 39c27eb..ac6cdfa 100644
--- a/src/fbd-feedback-manager.c
+++ b/src/fbd-feedback-manager.c
@@ -455,7 +455,7 @@ find_themefile (void)
       for (i = 0; i < g_strv_length (xdg_data_dirs); i++) {
         g_autofree gchar *config_path = NULL;
 
-        config_path = g_strconcat (xdg_data_dirs[i], "feedbackd/themes/", comp, ".json", NULL);
+        config_path = g_strconcat (xdg_data_dirs[i], "/feedbackd/themes/", comp, ".json", NULL);
         g_debug ("Searching for device specific themefile in %s", config_path);
 
         // Check if file exist
