From 340b738d7d4bb1c11a512d2f6489efc3c172648f Mon Sep 17 00:00:00 2001
From: Bhushan Shah <bhush94@gmail.com>
Date: Sat, 2 Jan 2021 09:06:22 +0530
Subject: [PATCH] voicecall: do not force the voice type

---
 drivers/qmimodem/voicecall.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/qmimodem/voicecall.c b/drivers/qmimodem/voicecall.c
index d6041a09..7e9e3e0b 100644
--- a/drivers/qmimodem/voicecall.c
+++ b/drivers/qmimodem/voicecall.c
@@ -259,7 +259,7 @@ static void dial(struct ofono_voicecall *vc, const struct ofono_phone_number *ph
 	memcpy(&vd->dialed, ph, sizeof(*ph));
 
 	arg.call_type_set = true;
-	arg.call_type = QMI_CALL_TYPE_VOICE_FORCE;
+	arg.call_type = QMI_CALL_TYPE_VOICE;
 
 	if (!qmi_voice_dial_call(
 				&arg,
-- 
2.29.2

