From: Jeremy Bicha <jeremy.bicha@canonical.com>
Date: Mon, 1 Aug 2022 10:05:56 -0400
Subject: utils: update for GNOME Desktop 43 thumbnail API changes

Fixes: #726
---
 src/chatty-utils.c | 21 +++++++++++++++++++++
 1 file changed, 21 insertions(+)

diff --git a/src/chatty-utils.c b/src/chatty-utils.c
index c95d938..6d38ad1 100644
--- a/src/chatty-utils.c
+++ b/src/chatty-utils.c
@@ -550,6 +550,26 @@ utils_create_thumbnail (GTask        *task,
     return;
   }
 
+#if defined(GNOME_DESKTOP_PLATFORM_VERSION) && GNOME_DESKTOP_PLATFORM_VERSION >= 43
+  thumbnail = gnome_desktop_thumbnail_factory_generate_thumbnail (factory, uri, content_type, NULL, &error);
+  if (!thumbnail) {
+    g_task_return_new_error (task, G_IO_ERROR, G_IO_ERROR_FAILED, "Failed to create thumbnail for file: %s (%s)", uri, error->message);
+
+    g_warning ("Failed to create thumbnail for file: %s", uri);
+
+    g_error_free (error);
+    return;
+  }
+
+  gnome_desktop_thumbnail_factory_save_thumbnail (factory, thumbnail, uri, mtime, NULL, &error);
+  if (error) {
+    g_task_return_new_error (task, G_IO_ERROR, G_IO_ERROR_FAILED, "Failed to create thumbnail for file: %s (%s)", uri, error->message);
+    g_error_free (error);
+    return;
+  }
+
+  g_task_return_boolean (task, TRUE);
+#else
   thumbnail = gnome_desktop_thumbnail_factory_generate_thumbnail (factory, uri, content_type);
 
   if (thumbnail) {
@@ -562,6 +582,7 @@ utils_create_thumbnail (GTask        *task,
   }
 
   g_task_return_boolean (task, TRUE);
+#endif
 }
 
 void
