From ad55246cef938caf1224ea202da35574a2c09fea Mon Sep 17 00:00:00 2001
From: Adrien Plazas <kekun.plazas@laposte.net>
Date: Tue, 3 Sep 2019 15:41:26 +0200
Subject: [PATCH 04/23] updates-page: Notify the stack when it needs attention

---
 src/gs-updates-page.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/src/gs-updates-page.c b/src/gs-updates-page.c
index 117e7bcd..95367040 100644
--- a/src/gs-updates-page.c
+++ b/src/gs-updates-page.c
@@ -295,6 +295,11 @@ refresh_headerbar_updates_counter (GsUpdatesPage *self)
 		gtk_style_context_add_class (gtk_widget_get_style_context (widget), "needs-attention");
 	else
 		gtk_style_context_remove_class (gtk_widget_get_style_context (widget), "needs-attention");
+
+	widget = GTK_WIDGET (gtk_builder_get_object (self->builder, "stack_main"));
+	gtk_container_child_set (GTK_CONTAINER (widget), GTK_WIDGET (self),
+				 "needs-attention", num_updates > 0,
+				 NULL);
 }
 
 static void
-- 
2.28.0

