From 691422199a9e85675c67737be197f6b0e67eef41 Mon Sep 17 00:00:00 2001
From: Adrien Plazas <kekun.plazas@laposte.net>
Date: Wed, 8 Sep 2021 14:44:05 +0200
Subject: [PATCH] icon: Ensure icon files exist

---
 lib/gs-appstream.c | 3 +++
 lib/gs-icon.c      | 9 +++++++++
 2 files changed, 12 insertions(+)

diff --git a/lib/gs-appstream.c b/lib/gs-appstream.c
index 756106078..8d53ed6a8 100644
--- a/lib/gs-appstream.c
+++ b/lib/gs-appstream.c
@@ -188,6 +188,9 @@ app_add_icon (GsApp  *app,
               AsIcon *as_icon)
 {
 	g_autoptr(GIcon) icon = gs_icon_new_for_appstream_icon (as_icon);
+	if (!icon)
+		return;
+
 	gs_app_add_icon (app, icon);
 }
 
diff --git a/lib/gs-icon.c b/lib/gs-icon.c
index 955ab0217..1b3927425 100644
--- a/lib/gs-icon.c
+++ b/lib/gs-icon.c
@@ -159,6 +159,9 @@ gs_icon_load_local (AsIcon *icon)
 		return NULL;
 
 	file = g_file_new_for_path (filename);
+	if (!g_file_query_exists (file, NULL))
+		return NULL;
+
 	return g_file_icon_new (file);
 }
 
@@ -184,6 +187,9 @@ gs_icon_load_remote (AsIcon *icon)
 	/* Load local files directly. */
 	if (g_str_has_prefix (url, "file:")) {
 		g_autoptr(GFile) file = g_file_new_for_path (url + strlen ("file:"));
+		if (!g_file_query_exists (file, NULL))
+			return NULL;
+
 		return g_file_icon_new (file);
 	}
 
@@ -237,6 +243,9 @@ gs_icon_load_cached (AsIcon *icon)
 	}
 
 	file = g_file_new_for_path (filename);
+	if (!g_file_query_exists (file, NULL))
+		return NULL;
+
 	return g_file_icon_new (file);
 }
 
-- 
2.30.2

