From 4e352f0f496cf15d946f6c6c58642675e6c6ef18 Mon Sep 17 00:00:00 2001
From: Adrien Plazas <kekun.plazas@laposte.net>
Date: Wed, 8 Sep 2021 17:29:17 +0200
Subject: [PATCH] app: Ensure themed icons exist

---
 lib/gs-app.c | 15 +++++++++++++--
 1 file changed, 13 insertions(+), 2 deletions(-)

diff --git a/lib/gs-app.c b/lib/gs-app.c
index 457f37c0e..b3d907e1e 100644
--- a/lib/gs-app.c
+++ b/lib/gs-app.c
@@ -1906,9 +1906,20 @@ gs_app_get_icon_for_size (GsApp       *app,
 	for (guint i = 0; priv->icons != NULL && i < priv->icons->len; i++) {
 		GIcon *icon = priv->icons->pdata[i];
 		guint icon_width = gs_icon_get_width (icon);
+		GtkIconTheme *icon_theme;
+		g_autoptr (GtkIconInfo) info = NULL;
+		GtkIconLookupFlags flags;
 
-		if (icon_width == 0 && G_IS_THEMED_ICON (icon))
-			return g_object_ref (icon);
+		if (icon_width != 0 || !G_IS_THEMED_ICON (icon))
+			continue;
+
+		icon_theme = gtk_icon_theme_get_default ();
+		flags = GTK_ICON_LOOKUP_FORCE_REGULAR;
+		info = gtk_icon_theme_lookup_by_gicon_for_scale (icon_theme, icon, size, scale, flags);
+		if (info == NULL)
+			continue;
+
+		return g_object_ref (icon);
 	}
 
 	if (scale > 1) {
-- 
2.30.2

