From 3b2469f07d600700d97114414d90731abe3a6cd5 Mon Sep 17 00:00:00 2001
From: Adrien Plazas <kekun.plazas@laposte.net>
Date: Fri, 3 Sep 2021 12:49:56 +0200
Subject: [PATCH] app: Be more strict with icon scales

---
 lib/gs-app.c | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/lib/gs-app.c b/lib/gs-app.c
index b3d907e1e..7eda3e668 100644
--- a/lib/gs-app.c
+++ b/lib/gs-app.c
@@ -1890,12 +1890,17 @@ gs_app_get_icon_for_size (GsApp       *app,
 		g_debug ("\tConsidering icon of type %s (%s), width %u×%u",
 			 G_OBJECT_TYPE_NAME (icon), icon_str, icon_width, icon_scale);
 
+		/* Ignore icons whose scale doesn't match, as GtkImage will fail
+		 * loading them. */
+		if (icon_scale != scale)
+			continue;
+
 		/* Ignore icons with unknown width and skip over ones which
 		 * are too small. */
-		if (icon_width == 0 || icon_width * icon_scale < size * scale)
+		if (icon_width == 0 || icon_width < size)
 			continue;
 
-		if (icon_width * icon_scale >= size * scale)
+		if (icon_width >= size)
 			return g_object_ref (icon);
 	}
 
-- 
2.30.2

