diff --git a/data/meson.build b/data/meson.build
index 2f42217..6bbd4d1 100644
--- a/data/meson.build
+++ b/data/meson.build
@@ -14,8 +14,7 @@ install_data(['postprocess.sh'],
   install_dir: get_option('datadir') / 'megapixels/',
   install_mode: 'rwxr-xr-x')
 
-settings_schemas = ['org.postmarketos.Megapixels.gschema.xml']
-schemas_dir = get_option('prefix') / get_option('datadir') / 'glib-2.0' / 'schemas'
-install_data(settings_schemas, install_dir: schemas_dir)
-gnome.compile_schemas(depend_files: files(settings_schemas))
-meson.add_install_script('glib-compile-schemas', schemas_dir)
+install_data(
+	'org.postmarketos.Megapixels.gschema.xml',
+	install_dir: join_paths (get_option('datadir'), 'glib-2.0', 'schemas')
+)
