From 02c61545d0e0ccca5a2d823c53d6ffdc86258b56 Mon Sep 17 00:00:00 2001
From: Martin Abente Lahaye <martin.abente.lahaye@gmail.com>
Date: Wed, 4 Aug 2021 18:22:57 -0400
Subject: [PATCH] trash: Handle permissions-restricted mounts

Closes #199
---
 src/trash.py | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/src/trash.py b/src/trash.py
index 7e6b012..4fd6519 100644
--- a/src/trash.py
+++ b/src/trash.py
@@ -88,6 +88,9 @@ def get_devices_trash(self):
 
         for mount in self._manager.get_mounts():
             mount_point = mount.get_root().get_path()
+            if not os.access(mount_point, os.W_OK):
+                continue
+
             path = os.path.join(mount_point, f".Trash-{os.getuid()}")
             trash.append((mount_point, path))
 
@@ -130,6 +133,10 @@ def list(self):
         return paths
 
     def trash(self, path):
+        # prevent shutil.move strange behavior
+        if not os.access(path, os.W_OK):
+            raise PermissionError(f"Can't access {path}")
+
         mount_point = utils.find_mount_point(path)
 
         trash_dir = self._trash.get(mount_point)
