From d649e112ae7fe2f02480e510444aac19dbc5b83f Mon Sep 17 00:00:00 2001
From: Arnaud Ferraris <arnaud.ferraris@collabora.com>
Date: Wed, 13 May 2020 14:12:13 +0200
Subject: [PATCH] driver: change small screen condition

---
 gnome-initial-setup/gis-driver.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/gnome-initial-setup/gis-driver.c b/gnome-initial-setup/gis-driver.c
index 14b62e0..624d241 100644
--- a/gnome-initial-setup/gis-driver.c
+++ b/gnome-initial-setup/gis-driver.c
@@ -616,7 +616,7 @@ monitor_is_small (GdkMonitor *monitor)
     return TRUE;
 
   gdk_monitor_get_geometry (monitor, &geom);
-  return geom.height < 800;
+  return geom.width < 800;
 }
 
 static void
-- 
2.26.2
