From 2091ac5f75a6e025be896c21179b04bc53ce4bb0 Mon Sep 17 00:00:00 2001
From: Arnaud Ferraris <arnaud.ferraris@gmail.com>
Date: Sat, 23 May 2020 17:14:36 +0200
Subject: [PATCH] mm-broadband-modem: improve voice capabilities detection

---
 src/mm-broadband-modem.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/mm-broadband-modem.c b/src/mm-broadband-modem.c
index fdbd13b0..6dadb90e 100644
--- a/src/mm-broadband-modem.c
+++ b/src/mm-broadband-modem.c
@@ -7400,7 +7400,7 @@ modem_voice_check_support (MMIfaceModemVoice *self,
 
     /* Check ATH support */
     mm_base_modem_at_command (MM_BASE_MODEM (self),
-                              "H",
+                              "I",
                               3,
                               FALSE,
                               (GAsyncReadyCallback)ath_format_check_ready,
-- 
2.26.2

