From a94d59407188a246bf07248f9570447008bf2e4b Mon Sep 17 00:00:00 2001
From: Arnaud Ferraris <arnaud.ferraris@collabora.com>
Date: Mon, 25 Oct 2021 14:32:04 +0200
Subject: [PATCH] build: fix build with meson 0.60.0

---
 data/appdata/meson.build | 1 -
 1 file changed, 1 deletion(-)

diff --git a/data/appdata/meson.build b/data/appdata/meson.build
index b5ee6540..200c3507 100644
--- a/data/appdata/meson.build
+++ b/data/appdata/meson.build
@@ -1,7 +1,6 @@
 appdata_conf = configuration_data()
 appdata_conf.set('appid', application_id)
 appdata = i18n.merge_file(
-  'appdata',
   input: configure_file(
     input: files('org.gnome.Calendar.appdata.xml.in.in'),
     output: 'org.gnome.Calendar.appdata.xml.in',
-- 
2.33.0

