From 43940ccd1f5976b54fc39b0882512815cdf6a0c3 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Guido=20G=C3=BCnther?= <agx@sigxcpu.org>
Date: Mon, 24 Feb 2020 13:00:59 +0100
Subject: [PATCH 01/16] timer: Use name that can be added to the event naming
 spec

---
 src/timer-face.vala | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/timer-face.vala b/src/timer-face.vala
index 5cad12f..8ee6732 100644
--- a/src/timer-face.vala
+++ b/src/timer-face.vala
@@ -71,7 +71,7 @@ public class Face : Gtk.Stack, Clocks.Clock {
             save ();
         });
 
-        bell = new Utils.Bell ("complete");
+        bell = new Utils.Bell ("timeout-completed");
         notification = new GLib.Notification (_("Time is up!"));
         notification.set_body (_("Timer countdown finished"));
         notification.set_priority (HIGH);
-- 
2.34.1

