From e88e14f4508b0365216fc081380352bb3006c976 Mon Sep 17 00:00:00 2001
From: Julian Sparber <julian@sparber.net>
Date: Mon, 14 Dec 2020 18:14:59 +0100
Subject: [PATCH 11/14] timer-setup: remove unused lables

---
 data/ui/timer-setup.ui | 22 ----------------------
 1 file changed, 22 deletions(-)

diff --git a/data/ui/timer-setup.ui b/data/ui/timer-setup.ui
index 8a2f3b1..588aedb 100644
--- a/data/ui/timer-setup.ui
+++ b/data/ui/timer-setup.ui
@@ -206,17 +206,6 @@
                     <property name="halign">center</property>
                     <property name="valign">center</property>
                     <property name="column_spacing">10</property>
-                    <child>
-                      <object class="GtkLabel">
-                        <property name="visible">True</property>
-                        <property name="can_focus">False</property>
-                        <property name="hexpand">True</property>
-                      </object>
-                      <packing>
-                        <property name="left_attach">0</property>
-                        <property name="top_attach">0</property>
-                      </packing>
-                    </child>
                     <child>
                       <object class="GtkSpinButton" id="h_spinbutton">
                         <property name="visible">True</property>
@@ -325,17 +314,6 @@
                         <property name="top_attach">0</property>
                       </packing>
                     </child>
-                    <child>
-                      <object class="GtkLabel">
-                        <property name="visible">True</property>
-                        <property name="can_focus">False</property>
-                        <property name="hexpand">True</property>
-                      </object>
-                      <packing>
-                        <property name="left_attach">6</property>
-                        <property name="top_attach">0</property>
-                      </packing>
-                    </child>
                   </object>
                   <packing>
                     <property name="expand">True</property>
-- 
2.34.1

