From 7403c9567f2f89fd588ad1796b1a816d5d12bccf Mon Sep 17 00:00:00 2001
From: Julian Sparber <julian@sparber.net>
Date: Tue, 15 Dec 2020 14:31:12 +0100
Subject: [PATCH 16/16] world-standalone: allow smaller window heights

---
 data/ui/world-standalone.ui | 14 --------------
 1 file changed, 14 deletions(-)

diff --git a/data/ui/world-standalone.ui b/data/ui/world-standalone.ui
index 46667d7..8f7acff 100644
--- a/data/ui/world-standalone.ui
+++ b/data/ui/world-standalone.ui
@@ -38,22 +38,8 @@
         <property name="can_focus">False</property>
         <property name="halign">center</property>
         <property name="valign">center</property>
-        <property name="margin_start">34</property>
-        <property name="margin_end">34</property>
-        <property name="margin_bottom">12</property>
         <property name="hexpand">True</property>
         <property name="vexpand">True</property>
-        <child>
-          <object class="GtkLabel" id="dummy_label">
-            <property name="visible">True</property>
-            <property name="can_focus">False</property>
-            <property name="label">  </property>
-          </object>
-          <packing>
-            <property name="left_attach">0</property>
-            <property name="top_attach">1</property>
-          </packing>
-        </child>
         <child>
           <object class="GtkLabel" id="time_label">
             <property name="visible">True</property>
-- 
2.34.1

